server/schedule/operator/builder.go

```diff
@@ -0,0 +1,638 @@
+		return p
+	}
+	return stepPlan{}
+}
+
+func (b *Builder) planReplace() stepPlan {
+	var best stepPlan
+	// add voter + remove voter.
+	for _, i := range b.toAdd.IDs() {
+		if add := b.toAdd.Get(i); !add.GetIsLearner() {
+			for _, j := range b.toRemove.IDs() {
+				remove := b.toRemove.Get(j)
+				if !remove.GetIsLearner() {
+					best = b.planReplaceLeaders(best, stepPlan{add: add, remove: remove})
+				}
+			}
+		}
+	}
+	// add learner + remove learner.
+	for _, i := range b.toAdd.IDs() {
+		if add := b.toAdd.Get(i); add.GetIsLearner() {
+			for _, j := range b.toRemove.IDs() {
+				remove := b.toRemove.Get(j)
+				if remove.GetIsLearner() {
+					best = b.planReplaceLeaders(best, stepPlan{add: add, remove: remove})
+				}
+			}
+		}
+	}
+	// promote learner + add learner + remove voter
+	for _, i := range b.toPromote.IDs() {
+		promote := b.toPromote.Get(i)
+		for _, j := range b.toAdd.IDs() {
+			if add := b.toAdd.Get(j); add.GetIsLearner() {
+				for _, k := range b.toRemove.IDs() {
+					if remove := b.toRemove.Get(k); !remove.GetIsLearner() && j != k {
+						best = b.planReplaceLeaders(best, stepPlan{promote: promote, add: add, remove: remove})
+					}
+				}
+			}
+		}
+	}
+	return best
+}
+
+func (b *Builder) planReplaceLeaders(best, next stepPlan) stepPlan {
+	// Brute force all possible leader combinations to find the best plan.
+	for _, leaderAdd := range b.currentPeers.IDs() {
+		if !b.allowLeader(b.currentPeers.Get(leaderAdd)) {
+			continue
+		}
+		next.leaderAdd = leaderAdd
+		for _, leaderRemove := range b.currentPeers.IDs() {
+			if b.allowLeader(b.currentPeers.Get(leaderRemove)) {
+				next.leaderRemove = leaderRemove
+				best = b.comparePlan(best, next)
+			}
+		}
+		if next.promote != nil {
+			next.leaderRemove = next.promote.GetStoreId()
+			best = b.comparePlan(best, next)
+		}
+		if next.add != nil && !next.add.GetIsLearner() {
+			next.leaderRemove = next.add.GetStoreId()
+			best = b.comparePlan(best, next)
+		}
+	}
+	return best
+}
+
+func (b *Builder) planPromotePeer() stepPlan {
+	for _, i := range b.toPromote.IDs() {
+		p := b.toPromote.Get(i)
+		return stepPlan{promote: p}
+	}
+	return stepPlan{}
+}
+
+func (b *Builder) planRemovePeer() stepPlan {
+	var best stepPlan
+	for _, i := range b.toRemove.IDs() {
+		r := b.toRemove.Get(i)
+		for _, leader := range b.currentPeers.IDs() {
+			if b.allowLeader(b.currentPeers.Get(leader)) {
+				best = b.comparePlan(best, stepPlan{remove: r, leaderRemove: leader})
+			}
+		}
+	}
+	return best
+}
+
+func (b *Builder) planAddPeer() stepPlan {
+	var best stepPlan
+	for _, i := range b.toAdd.IDs() {
+		a := b.toAdd.Get(i)
+		for _, leader := range b.currentPeers.IDs() {
+			if b.allowLeader(b.currentPeers.Get(leader)) {
+				best = b.comparePlan(best, stepPlan{add: a, leaderAdd: leader})
+			}
+		}
+	}
+	return best
+}
+
+// Pick the better plan from 2 candidates.
+func (b *Builder) comparePlan(best, next stepPlan) stepPlan {
+	if best.empty() {
```

Comment:
If the builder do:
```
 b. AddPeer({StoreID: 5}). RemovePeer({StoreID: 4})
```
store 4 is a leader peer, store 5 not allow leader. the best plan should the empty plan or report an error.

Reply:
It makes sense. I'll do it.